pkg/menu/menu.go

```diff
@@ -0,0 +1,72 @@
+package menu
+
+import (
+	"io"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/gizak/termui/v3/widgets"
+)
+
+const menuWidth = 50
+
+// newParagraph will return a paragraph object with given initial text.
+func newParagraph(initText string, border bool, location int, wid int, ht int) *widgets.Paragraph {
+	p := widgets.NewParagraph()
+	p.Text = initText
+	p.Border = border
+	p.SetRect(0, location, wid, location+ht)
+	p.TextStyle.Fg = ui.ColorWhite
+	ui.Render(p)
+	return p
+}
+
+func processInput(input, warning *widgets.Paragraph, isValid func(string) (bool, string), uiEvents <-chan ui.Event) (string, error) {
+	// keep tracking all input from user
+	for {
+		e := <-uiEvents
+		if e.Type != ui.KeyboardEvent {
+			continue
+		}
+		switch e.ID {
+		// if user input control-d, raise error io.EOF meaning no more input.
+		case "<C-d>":
+			return "", io.EOF
+		case "<Enter>":
+			if isValid == nil {
+				return input.Text, nil
+			}
+			valid, warningWords := isValid(input.Text)
+			if valid {
+				return input.Text, nil
+			}
+			input.Text = ""
+			warning.Text = warningWords
+			ui.Render(input)
+			ui.Render(warning)
+		// If user hit backspace, remove the last input character
+		case "<Backspace>":
+			input.Text = input.Text[:len(input.Text)-1]
+			ui.Render(input)
+		default:
+			if warning.Text != "" {
+				warning.Text = ""
+				ui.Render(warning)
```

Comment:
I am not sure I understand this. Do you want to render the warning before you set it to ""?

Reply:
Is there a race condition between using the warning.Text and setting it? You may want to run the go race detector.

Reply:
I think render is after I set warning to "" since in my mind render is to implement the change. and what you mean about  a race condition ? what I want to do here is to reset the warning when user begin to write a new input. If I set the warning.Text and then render it, will there be a race condition?

Reply:
ah ha. I did not understand that this was clearing the message :-)
I think there's likely no problem with races but just wondered.